hack/build-and-push.sh

```diff
@@ -133,95 +180,324 @@ if [ "$SKIP_BUILD" == "" ]; then
+        if [ -n "$digest" ]; then
+            cache_set "$image" "$digest"
         fi
-        if [ -f "${task_dir}/TROUBLESHOOTING.md" ]; then
-            ANNOTATIONS+=("dev.tekton.docs.troubleshooting=${VCS_URL}/tree/${VCS_REF}/${task_dir}TROUBLESHOOTING.md")
+    fi
+    echo "$digest"
+}
+
+# Attach migration file to given task bundle.
+# Arguments: task_dir, concrete_task_version, task_bundle
+attach_migration_file() {
+    local -r task_dir=$1
+    local -r concrete_task_version=$2
+    local -r task_bundle=$3
+
+    local -r task_file="${task_dir}/${task_name}.yaml"
+    local -r migration_file="${task_dir}/migrations/${concrete_task_version}.sh"
+
+    if [ ! -e "$migration_file" ]; then
+        return 0
+    fi
+
+    # Check if task bundle has an attached migration file.
+    local filename
+    local found=
+    local artifact_refs
+
+    # List attached artifacts, that have specific artifact type and annotation.
+    # Then, find out the migration artifact.
+    artifact_refs=$(
+        retry oras discover "$task_bundle" --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" --format json | \
+        jq -r "
+            .manifests[]
+            | select(.annotations.\"${ANNOTATION_TASK_MIGRATION}\" == \"true\")
+            | .reference"
+    )
+    while read -r artifact_ref; do
+        if [ -z "$artifact_ref" ]; then
+            continue
         fi
-        if [ -f "${task_dir}/USAGE.md" ]; then
-            ANNOTATIONS+=("dev.tekton.docs.usage=${VCS_URL}/tree/${VCS_REF}/${task_dir}USAGE.md")
+        filename=$(
+            retry oras pull --format json "$artifact_ref" | jq -r "
+                .files[]
+                | select(.annotations.\"org.opencontainers.image.title\" == \"${concrete_task_version}.sh\")
+                | .annotations.\"org.opencontainers.image.title\"
+                "
+        )
+
+        if [ -n "$filename" ]; then
+            if diff "$filename" "$migration_file" >/dev/null; then
+                found=true
+                break
+            else
+                echo "error: task bundle $task_bundle has migration artifact $artifact_ref, but the migration content is different: $filename" 1>&2
+                exit 1
+            fi
         fi
+    done <<<"$artifact_refs"
 
-        ANNOTATION_FLAGS=()
-        for annotation in "${ANNOTATIONS[@]}"; do
-            ANNOTATION_FLAGS+=("--annotate" "$(escape_tkn_bundle_arg "$annotation")")
-        done
+    if [ -n "$found" ]; then
+        return 0
+    fi
 
-        output=$(tkn_bundle_push "${ANNOTATION_FLAGS[@]}" -f "$prepared_task_file" "$task_bundle" | save_ref "$task_bundle" "$OUTPUT_TASK_BUNDLE_LIST")
-        echo "$output"
-        task_bundle_with_digest="${output##*$'\n'}"
+    (
+        cd "${migration_file%/*}"
+        retry oras attach \
+            --registry-config "$AUTH_JSON" \
+            --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" \
+            --distribution-spec v1.1-referrers-tag \
```

Comment:
Do we need to specifically use the referrers-tag rather than the referrers-api?

Reply:
:thinking:  Is referrers-api preferred?

Reply:
I would say so - if I understand correctly, referrers-tag should be the fallback for registries that don't support referrers-api

Can we leave it unspecified and let oras decide which distribution spec to use?

Reply:
Can `oras attach` determine that by itself?

`--distribution-spec` is specified in some tasks explicitly.

Reply:
According to c746a251af63d1b91bf79b8f759737aaba0227ab, it can

Reply:
It's removed from the script.